TFNetworkLayer.py

```diff
@@ -1043,53 +1047,238 @@ def batch_norm(self, data,
-            initial_value=tf.zeros(data.get_bc_spatial_batch_shape()),
+            initial_value=tf.zeros(variables_shape),
             name="%s_%s_mean" % (self.name, data.name),
             trainable=False))
         # Use exponential moving average of batch mean.
         # Note: We could also use cumulative moving average. Our Theano implementation does that for inference.
         sample_mean = tf.assign_add(sample_mean, (mean - sample_mean) * momentum)
+      else:
+        sample_mean = tf.reshape(sample_mean, shape=variables_shape)
       if sample_variance is None:
         # Note: Our Theano implementation does not use a moving average for this.
         with self.var_creation_scope():
           sample_variance = self.add_param(tf.Variable(
-            initial_value=tf.ones(data.get_bc_spatial_batch_shape()),
+            initial_value=tf.ones(variables_shape),
             name="%s_%s_variance" % (self.name, data.name),
             trainable=False))
         sample_variance = tf.assign_add(sample_variance, (variance - sample_variance) * momentum)
+      else:
+        sample_variance = tf.reshape(sample_variance, shape=variables_shape)
       # If train or if force_sample, use default use_sample=0.0, otherwise use_sample=1.0.
       use_sample = 1.0 + tf.cast(tf.logical_or(self.network.train_flag, force_sample), tf.float32) * (use_sample - 1.0)
       mean = (1. - use_sample) * mean + use_sample * sample_mean
       variance = (1. - use_sample) * variance + use_sample * sample_variance
-      bn = (data.placeholder - mean) * tf.rsqrt(variance + epsilon)
+      bn = (input_data.placeholder - mean) * tf.rsqrt(variance + epsilon)
       if use_std:
         if gamma is None:
           with self.var_creation_scope():
             gamma = self.add_param(tf.Variable(
-              initial_value=tf.ones(data.get_bc_spatial_batch_shape()),
+              initial_value=tf.ones(variables_shape),
               name="%s_%s_gamma" % (self.name, data.name),
               trainable=True))
         bn *= gamma
       if use_shift:
         if beta is None:
           with self.var_creation_scope():
             beta = self.add_param(tf.Variable(
-              initial_value=tf.zeros(data.get_bc_spatial_batch_shape()),
+              initial_value=tf.zeros(variables_shape),
               name="%s_%s_beta" % (self.name, data.name),
               trainable=True))
         bn += beta
+      if data.is_batch_feature_major and not masked_time:
+        # Transpose back to NCHW
+        perm = data.get_axes()
+        perm = perm[:1] + perm[-1:] + perm[1:-1]
+        bn = tf.transpose(bn, perm=perm)
       return bn
 
+  def fused_batch_norm(self, data,
+                       axes='f',
+                       use_shift=True, use_std=True,
+                       momentum=0.99, epsilon=1e-3,
+                       sample_mean=None, sample_variance=None,
+                       gamma=None, beta=None,
+                       masked_time=True):
```

Comment:
Please remove the default values. They are all already given by `batch_norm`, and you should never call the function directly.